Add set_number_of_grid_points utility

wthrowe opened the pull request:
## Proposed changes

This allows resizing existing objects in a similar way to the sizing in `make_with_value`.  It fixes one case where I was copying data just to get the correct structure, and will likely be convenient in other cases.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
Replace `destructive_resize_components` with the new `set_number_of_grid_points`.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Test make_with_value in appropriate files

Move the tests for DataVector and Variables from
Test_MakeWithValue.cpp to the tests for the classes, since that is
where the impl being tested is defined.  Tensor was already tested.

Also add an independent test for SpinWeighted, although it was already
tested in combination with the vector classes.

wthrowe committed:
Allow array, TaggedTuple as make_with_value input

wthrowe committed:
Add set_number_of_grid_points

Similar to the sizing behavior of make_with_value, but without the
making or the value.

wthrowe committed:
Add set_number_of_grid_points implementations

wthrowe committed:
Use set_number_of_grid_points in UpdateU

nilsdeppe commented on src/Utilities/GenericResize.hpp:
```diff
@@ -0,0 +1,88 @@
+///
+/// \see GenericResizeImpls, make_with_value
+template <typename T, typename U>
+void generic_resize(const gsl::not_null<T*> result, const U& pattern) {
```
I think GCC is unhappy because the arguments may be unused according to it...

nilsdeppe reviewed (commented):
LGTM, I think just a minor thing to get GCC to be happy. Thanks for doing this!

knelli2 commented on src/Utilities/GenericResize.hpp:
```diff
@@ -0,0 +1,88 @@
+/// Implementations of \link generic_resize \endlink
+///
+/// Specializations may be trivial, meaning the object is similar to a
+/// `double` and doesn't actually represent grid data.  The only
+/// requirement for a trivial specializations is defining `is_trivial`
```
`for trivial specializations`

knelli2 commented on src/Utilities/GenericResize.hpp:
```diff
@@ -0,0 +1,88 @@
+  }
+}
+
+template <>
+struct GenericResizeImpls::GenericResizeImpl<double> {
```
Maybe also add a specialization for `std::vector`? I think that, along with the implementations in the next commit, covers all the common containers we'd normally use.

wthrowe replied:
I added the same cases that work for `make_with_value`, which I think were chosen as types with arithmetic operations.

That said, I can't think of any reason to restrict these utilities to such types, and supporting `std::vector` sounds potentially useful, so I'll add it (and also for `make_with_value`).

knelli2 reviewed (changes requested):
Just a couple small things. If you decide to add the `std::vector` impl, just squash.

wthrowe reviewed (commented):
Based mostly on conversations elsewhere on comparisons with the "destructive_resize" family of functions, I'm going to rename this.  Particularly with potential support for `std::vector`, I've decided I don't like the name "resize" because it acts on the elements of containers in some cases and sometimes on the object itself.  Instead, I'm going to go with something like `set_number_of_grid_points`, since that's pretty unambiguous.

The existing function `destructive_resize_components` is a special case of this function, and can be removed in favor of it (except for the unused three-argument form).

wthrowe committed:
Implement set_number_of_grid_points(std::vector)

wthrowe committed:
Remove LIKELY markings on constant expressions

wthrowe committed:
Remove destructive_resize_components

Replace with set_number_of_grid_points or remove it where it was
unnecessary.  Also replaced a few manual resizings I happened to see.

A few cases were resizing to the wrong size or immediately discarding
the allocation, so those should be slightly more efficient now.

wthrowe committed:
Allow make_with_value size from reference_wrapper

wthrowe commented:
Renamed.  Replaced `destructive_resize_components` since it is effectively the same.  Did a few other things required to make that work that you can see in the history.  Removed some silly branch-prediction hints that I happened to notice.

A few functions ended up with unused arguments that had previously only been used for unnecessary sizing.  I didn't investigate in depth because this PR already touches enough random code, but I found it odd.  In case anyone is interested, they were (specific overloads of):
```
src/Domain/FaceNormal.cpp unnormalized_face_normal
src/Evolution/Systems/Cce/BoundaryData.cpp du_j_worldtube_data
src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp partial_derivative
```
I suspect other overloads of those functions probably still have similar effectively unused arguments, because I only removed sizing calls that were very obviously unneeded.

nilsvu commented on src/Utilities/MakeWithValue.hpp:
```diff
@@ -81,7 +85,7 @@ struct MakeWithValueImpl {
 /// `double`.
 ///
-/// \see MakeWithValueImpls
+/// \see MakeWithValueImpls, generic_resize
```
`set_number_of_grid_points`

wthrowe replied:
Fixed when I next push.

nilsvu commented on src/Domain/FaceNormal.cpp:
```diff
@@ -10,21 +10,19 @@
 void unnormalized_face_normal(
     const gsl::not_null<tnsr::i<DataVector, VolumeDim, TargetFrame>*> result,
-    const Mesh<VolumeDim - 1>& interface_mesh,
+    const Mesh<VolumeDim - 1>& /*interface_mesh*/,
```
Could you add a note that this is left over from a refactor and can be deleted?

wthrowe replied:
Filed #5177.

nilsvu reviewed (dismissed):
Amazing! Thanks for doing this!

The two small changes below can happen in a follow-up to avoid rebasing this PR.

nilsvu commented:
CI has a linking issue:

```
/work/build/bin/RunTests: symbol lookup error: /work/build/lib/libTest_Domain.so: undefined symbol: _ZN6domain19flat_logical_metricILm1EEEvN3gsl8not_nullIP6TensorI10DataVectorN7brigand4listIJNS5_17integral_constantIiLi1EEES8_EEENS6_IJN13Tensor_detail15TensorIndexTypeIXT_EL4UpLo1EN5Frame14ElementLogicalEL9IndexType0EEESG_EEEEEERKS3_IS4_NS6_IJNS7_IiLi2EEES8_EEENS6_IJNSB_IXT_ELSC_0ENSD_8InertialELSF_0EEESG_EEEE
```

knelli2 reviewed (commented):
Looks good except for the linking issue.

Question @wthrowe: In some places, you completely remove a call to `destructive_resize_components` instead of replacing it with `set_number_of_grid_points`. Why is that?

wthrowe commented:
Added some explicit instantiations to hopefully fix the linking error.  I don't get it locally, so I'l have to wait for CI.  Also rebased on develop.

> Question @wthrowe: In some places, you completely remove a call to `destructive_resize_components` instead of replacing it with `set_number_of_grid_points`. Why is that?

For most of them the next thing done with the variable was obviously something that would check the size and resize if necessary, like assigning from a blaze expression template.  For a few the next thing was an assignment from an rvalue, in which case the resize was actually harmful.  I suspect some of the remaining ones are also unnecessary, but I wasn't checking anything where it wasn't obvious.

knelli2 reviewed (approved):
The singular clang-tidy error doesn't seem important enough to hold up this PR.

nilsdeppe commented:
CI failures are timeouts and unrelated clang-tidy complaints.

Merging now to minimize conflicts of this PR.

nilsvu commented:
No, it also wasn't introduced in this PR (edit: @nilsdeppe was faster)

nilsvu commented:
Added upgrade instructions. @wthrowe feel free to edit.

wthrowe commented:
Removed upgrade instructions.  After discussion elsewhere I still don't agree that they are appropriate.
